### PR TITLE
add test for Issue 15056 - Unstored RAII struct return yields bogus error

### PR DESCRIPTION
**Pull request by MartinNowak**

### CODE REVIEW

**Comment by MartinNowak** [comment]
Fixed by reverting part of #5003.

**Comment by MartinNowak** [comment]
This is blocking 2.068.2, a proper fix to 14708 is too complex, it requires to lift the destruction of temporaries from the IR gen to the semantic phase so we can properly handle nothrow code.
Another reason to revert the fix is that it currently always generates try/finally for any temporary with destructor, this is a lot slower for nothrow code.

File: test/compilable/test15056.d [hunk 1]
```diff
@@ -0,0 +1,27 @@
+version (Windows)
+{
+    version (LP_64)
+        import core.stdc.stdlib;
+    else
+        void* alloca(size_t) { return null; } // doesn't currently work b/c SEH remains present in nothrow code
```
Comment (9rnsr) [inline test/compilable/test15056.d]: From the auto-tester error, isn't this `nothrow`?

**Comment by 9rnsr** [comment]
https://issues.dlang.org/show_bug.cgi?id=15056

**Comment by 9rnsr** [comment]
I can accept this, but I don't know whether the issue 15056 case had really worked correctly with 2.067 or before.

Assign to @WalterBright

**Comment by 9rnsr** [comment]
And, this change will reopen issue 14708 again. Maybe we need to document it in the changelog?

**Comment by MartinNowak** [comment]
> I can accept this, but I don't know whether the issue 15056 case had really worked correctly with 2.067 or before.

It is valid code and worked correctly w/ nothrow. It partly worked for throwing code as well b/c of 14708.

> And, this change will reopen issue 14708 again. Maybe we need to document it in the changelog?

Done.

**Comment by 9rnsr** [comment]
> It is valid code and worked correctly w/ nothrow.

Even in `nothrow` function, `Throwable` (especially `Error`) object may be thrown and destructors need to be called because they're in `finally` block. That's why EH had been added unconditionally around any temporaries.

**Comment by MartinNowak** [comment]
> Even in nothrow function, Throwable (especially Error) object may be thrown and destructors need to be called because they're in finally block.

We don't do that for try/catch (see http://dpaste.dzfl.pl/5c4848b6a48b or [Issue 7018](https://issues.dlang.org/show_bug.cgi?id=7018)) and it'd be expensive w/ dmd's horrible EH code.
Throwing Errors in nothrow functions always had the fishy "it'll do some cleanup" semantic.

**Comment by WalterBright** [comment]
Seems like we need to rethink this. It should not be causing regressions and inefficiencies. I agree with the reversion.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Comment by mleise** [comment]
Sorry for the trouble. Dlang may be missing a specification part about alloca() and some test cases along with that. It is one of these little systems programming tools that can give an edge over higher level languages, but breaks assumptions about the stack layout in compilers and requires an amount of support code that makes you wonder if it is worth it.
I agree with Martin that thrown Errors mean that the program is in an inconsistent state and needs to leave at the next exit. It may even be written in the spec, that cleanup can be omitted on errors. Maybe bending the rules a bit, LDC and GDC could even apply their 'nothrow' back-end attributes. I don't know if that optimization really buys us anything, but I heard @ibuclaw talk about it at least and there is posts like this https://www.bountysource.com/issues/1407205-arm-runnable-opover2-d-fails-xopequals-fallback-aborts-instead-of-throwing-exception
Neither GDC nor LDC will make the first step, put a nail in the coffin and declare Errors unrecoverable, because folks would say their compiler is kaput. Meantime people like myself take that undefined state on Errors idea more literal and think that alloca() should work in nothrow functions.
So here is my proposal for a new nothrowsatall attribute … (just kidding)

**Pull request merged**